### PR TITLE
Refactor user flows to use nextjs page structure

### PR DESCRIPTION
**Pull request by soerenbf**
## Purpose

Refactor the codebase to make it simpler to understand, and follow nextJS and react best practices. This also improves both deposit/withdrawal flows by making them consistent, and makes it possible to access withdrawal/deposit details (progress page) after submitting from the transaction history.

## Changes

- Refactor monolith template to separate pages
- Link to transaction progress page from transaction history
- Remove dead code

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: frontend/src/contracts/use-generate-contract.ts [hunk 1]
```diff
@@ -4,9 +4,11 @@ import { toWei } from "../helpers/number";
 import useWallet from "../hooks/use-wallet";
 import MOCK_ABI from "./abis/MOCK_ABI.json";
 
+const ERC20_ALLOWANCE = "1000000000000";
+
 const useGenerateContract = (address: string, enabled: boolean) => {
     const { context } = useWallet();
-    const amountToApprove = toWei("1000000000000");
+    const amountToApprove = toWei(ERC20_ALLOWANCE); // TODO: config.
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/contracts/use-generate-contract.ts:11]: Yes, meant "make this configurable from environment", but for now we'll just keep it as is I think

[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/contracts/use-generate-contract.ts:11]: Will remove todo for now.

File: frontend/pages/deposit/overview.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import { Components } from "src/api-query/__generated__/AxiosClient";
+import { ContractTransaction, errors } from "ethers";
+import addresses from "@config/addresses";
+import useRootManagerContract from "src/contracts/use-root-manager";
+import { routes } from "src/constants/routes";
+import useGenerateContract from "src/contracts/use-generate-contract";
+import { useTransactionFlowStore } from "src/store/transaction-flow";
+import { useRouter } from "next/router";
+
+const WithdrawOverview: NextPage = () => {
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/deposit/overview.tsx:12]: You're not crazy!

File: frontend/src/helpers/detectCcdProvider.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { WalletApi } from "@concordium/browser-wallet-api-helpers";
 
-export async function detectCcdProvider(timeout = 5000): Promise<WalletApi> {
+async function detectCcdProvider(timeout = 5000): Promise<WalletApi> {
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/helpers/detectCcdProvider.ts:3]: Nope..

File: frontend/src/contracts/use-ccd-contract.ts [hunk 4]
```diff
@@ -101,22 +102,11 @@ const useCCDContract = (ccdAccount: string | null, enabled: boolean) => {
...
         }
-        // if (!energy) {
-        //   throw new Error("Energy is undefined");
-        // }
-
-        // const maxContractExecutionEnergy = BigInt(
-        //   Math.ceil((energy * 10 ** (ccdToken.decimals / 2)) / 100) * 100
-        // );
 
         const maxContractExecutionEnergy = BigInt(30000);
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/contracts/use-ccd-contract.ts:106]: Yes, this should change in the PR for bigint support

File: frontend/src/components/templates/transfer-progress/TransferProgress.tsx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+                        >
+                            {props.isWithdraw
+                                ? step > 1
+                                    ? "Withdraw Processed!"
+                                    : props.canWithdraw
+                                    ? "Your withdraw is ready for approval."
+                                    : "Your withdraw is in progress. Please come back later."
+                                : step > 1
+                                ? "Deposit Processed!"
+                                : "Your deposit is in progress."}
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/components/templates/transfer-progress/TransferProgress.tsx:232]: Removed the nested ternary expressions.

File: frontend/src/api-query/use-watch-withdraw/useWatchWithdraw.ts [hunk 6]
```diff
@@ -1,11 +1,17 @@
...
 interface Params extends Paths.WatchWithdrawTx.PathParameters {}
+type Options = UseQueryOptions<
+    Components.Schemas.WatchWithdrawalResponse,
+    unknown,
+    Components.Schemas.WatchWithdrawalResponse,
+    (string | Params | undefined)[]
+>;
 
-const useWatchWithdraw = (params?: Params, options?: any) => {
+const useWatchWithdraw = (params?: Params, options?: Options) => {
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/api-query/use-watch-withdraw/useWatchWithdraw.ts:14]: Yes good point.. didn't give this much thought when going though this tbh. Will clean it up.

File: frontend/src/api-query/use-eth-merkle-proof/useEthMerkleProof.ts [hunk 7]
```diff
@@ -14,11 +17,21 @@ const useEthMerkleProof = (params: Params, options?: any) => {
             const client = await getClient();
 
             if (!client) throw new Error("Client not initialized.");
+            if (params.event_id === undefined || params.tx_hash === undefined)
+                throw new Error("Should not be enabled with partial params");
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/api-query/use-eth-merkle-proof/useEthMerkleProof.ts:21]: yes, the rule that enforces blocks is not enabled though..

File: frontend/pages/history/[direction].tsx [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import { useEffect } from "react";
+
+const TransferHistory: NextPage = () => {
+    const { query, prefetch, isReady } = useRouter() as QueryRouter<{ direction: BridgeDirection }>;
+
+    useEffect(() => {
+        if (isReady) {
+            prefetch(
+                routes.history(
+                    query.direction === BridgeDirection.Deposit ? BridgeDirection.Withdraw : BridgeDirection.Deposit
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/history/[direction].tsx:16]: We expect the user to want to navigate from deposit to withdraw and vice versa. Prefetch is a performance optimization to make sure the next route the user wants to go to from the current page is available as fast as possible.

File: frontend/pages/deposit/[tx].tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { NextPage } from "next";
+import TransferProgress from "@components/templates/transfer-progress";
+import useWatchDeposit from "src/api-query/use-watch-deposit/useWatchDeposit";
+import { useRouter } from "next/router";
+import { QueryRouter } from "src/types/config";
+import { routes } from "src/constants/routes";
+import { useEffect } from "react";
+
+/** Interval in ms for how often to query for deposit status */
+const QUERY_INTERVAL = 10000;
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/deposit/[tx].tsx:10]: Will give this a look

File: frontend/src/components/templates/history/History.tsx [hunk 10]
```diff
@@ -351,17 +286,16 @@ const History: React.FC<Props> = ({}) => {
...
-                                                                fontColor={
-                                                                    transaction.Withdraw.tx_hash ? "Black" : "Yellow"
-                                                                }
+                                                                fontColor={ethHash ? "Black" : "Yellow"}
                                                             >
-                                                                {transaction.Withdraw.tx_hash ? (
+                                                                {ethHash ? (
                                                                     <a
-                                                                        href={`https://goerli.etherscan.io/tx/${transaction.Withdraw.tx_hash}`}
+                                                                        href={ethTransactionUrl(ethHash)}
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/components/templates/history/History.tsx:293]: Yes, I noticed as well. I changed it to use the environment variable `NEXT_PUBLIC_ETHEREUM_PROVIDER_NETWORK`, which was probably the original intention of this anyway.

File: frontend/src/components/templates/history/History.tsx [hunk 11]
```diff
@@ -95,9 +96,14 @@ const History: React.FC<Props> = ({}) => {
         });
     }, []);
 
-    if (!context.account) {
-        return <></>;
-    }
+    useEffect(() => {
+        // Effects only run client-side, nextJS router is only available on the client.
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/components/templates/history/History.tsx:100]: The server side render fails if the code in the `useEffect` hook runs, so it's merely to explain that the reason why this is included in a `useEffect` is to postpone running it until the first client side render pass

File: frontend/src/components/templates/history/History.tsx [hunk 12]
```diff
@@ -25,54 +28,52 @@ import {
...
-            setOpen(index);
-        }
+
+        const route = isDeposit(transaction) ? routes.deposit.tx(txHash) : routes.withdraw.tx(txHash);
+        push(route);
     };
 
+    const linkClick: MouseEventHandler = (e) => {
+        e.stopPropagation();
+    };
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/components/templates/history/History.tsx:61]: What? Are you commenting on the fact that there is a function dedicated to this single purpose, or what does your comment mean?

File: frontend/src/components/templates/history/History.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/components/templates/history/History.tsx:100]: I made the comment more descriptive

File: frontend/src/api-query/use-wallet-transactions/useWalletTransactions.ts [hunk 13]
```diff
@@ -1,23 +1,51 @@
+import useWallet from "@hooks/use-wallet";
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/api-query/use-wallet-transactions/useWalletTransactions.ts]: Changed this to `useEthWallet` from `use-eth-wallet`!

File: frontend/src/components/templates/history/History.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (shjortConcordium) [reply frontend/src/components/templates/history/History.tsx:61]: I guess I meant the function doesn't need to be inside the component.

File: frontend/pages/deposit/[tx].tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/deposit/[tx].tsx:10]: I gathered all the api query hooks in a single file and aligned the interval constants (also present in the same file).

File: frontend/src/helpers/price-usd.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+export const getPrice = async (token: "ETH" | "CCD"): Promise<number> => {
+    if (token === "ETH") {
+        const ethResponse = await axios.get(
+            "https://api.coingecko.com/api/v3/simple/price?ids=ethereum&vs_currencies=usd"
+        );
+        return ethResponse.data.ethereum.usd;
+    } else {
+        const ethResponse = await axios.get(
+            "https://api.coingecko.com/api/v3/simple/price?ids=concordium&vs_currencies=usd"
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/helpers/price-usd.ts:11]: I don't know what lies underneath the original decision to get prices from coingecko, but I think we should just go with this for now as a starting point.

[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/src/helpers/price-usd.ts:11]: As for the config file part of it, the issue is that we can't ensure the endpoints return the same thing anyway, so I think a scoped implementation such as this makes more sense. In the future, we could add support for more price providers, but I think we'd have to implement a separate price function for each anyway, and then make it configurable which of them to choose.

File: frontend/pages/deposit/overview.tsx [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        } catch (error) {
+            // TODO: remove...
+            console.error("gas reason:", error);
+
+            // else, the user did not approve or doesn't have enought tokens and we see the error
+            if (error?.reason) {
+                throw new Error(error?.reason);
+            } else {
+                throw error;
+            }
```
[the remark this answers is not in the document]
Comment (shjortConcordium) [reply frontend/pages/deposit/overview.tsx]: Is there a reason you ignored my great suggestion?

**Review by shjortConcordium — APPROVED** [review]

File: frontend/pages/deposit/overview.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/deposit/overview.tsx]: Wait... I thought I did this locally 😄 I meant to at least..

[the remark this answers is not in the document]
Comment (soerenbf) [reply frontend/pages/deposit/overview.tsx]: It's done...

**Pull request merged**